.github/workflows/publish.yml

```diff
@@ -0,0 +1,52 @@
+name: Publish
+
+on:
+  release:
+    types: [published]
```

Comment:
This is the last of my concerns. Having this inverse trigger can lead to weird sync issues.
For example, last week we were releasing `pip-tools` where this trigger is still used. And we ended up with a Git tag in repo and a GitHub Release existing for like 15-20 hours with no PyPI release, due to some approval challenges. Having a `workflow_dispatch` can prevent it when the sequence of actions performed is different.

Reply:
I assume you want the workflow to instead be:

* Push to `main`
* Publish to PyPI (via manually triggering publish GitHub workflow)
* Create and push Git tag
* Create GitHub release

How do you prevent accidental triggering of publish GitHub workflow? A solution here is to prevent upload except when the wheel's version is of the form "X.Y.Z".

What if there's a problem during tag or release creation, where you have a version on PyPI with no associated tag? A solution here is to yank, but in my opinion this is a much worse outcome than the problems you mentioned, as these can be pinned by downstream users.

Reply:
As I said [above](https://github.com/pypa/packaging/pull/893#issuecomment-2862480130):

> I don't have familiarity with using the `workflow_dispatch` GitHub workflows event, and don't feel comfortable switching to that workflow. If you want it, could you please make changes to this PR (or submit a new PR targeting this branch)?

I've never used `workflow_dispatch` for package release, so I'm not confident enough to write workflow definitions using it.

Reply:
@webknjaz yes please

Reply:
I've encountered a significant issue with using `workflow_dispatch` to publish the workflow. In [cryptography's publish workflow](https://github.com/pyca/cryptography/blob/ddf97733f902900e2d2333dafa24f4f49c26285e/.github/workflows/pypi-publish.yml), the `workflow_dispatch` must be run from `main`, but the wheel can be built from any random branch/tag, as there's no hard linking from the ref the publish workflow is run on. In `cryptography`'s case, the publish workflow uses a custom download-artifact action, which makes it harder to follow.

This break in the provenance chain, and I cannot rely on an artefact being generated by a GitHub workflow on a specific Git commit.

@webknjaz is there a way to better link the publish workflow to the build workflow?

Reply:
Build will be able to provide it directly in the next release as JSON. Not sure if that helps, as you'd still need to parse the JSON, but if that's easy, it might be simpler.